### PR TITLE
Add syslog support to ni_grpc_device_server

### PR DESCRIPTION
**Pull request by zhindes**
### What does this Pull Request accomplish?

This adds syslog support to ni_grpc_device_server. It does this by:
* Adding a simple logging interface to ni_grpc_device_server.
   * This should be used instead of direct stdout/stderr access.
   * The implementation uses a simple va_list-function that can be implemented for other logging.
   * There is a default stdout/stderr implementation
* Implemented a syslog logging function available on linux
* Added _slightly_ more robust command line option parsing with support for `--use-syslog`. I expect this may be replaced with a more generic daemon option in the future.

### Why should this Pull Request be merged?

This is early work towards making ni_grpc_device_server natively-daemonizable for on-boot configuration.

### What testing has been done?

Normal invocation and sad/happy paths:
```
admin@NI-PXIe-8880-03095FC1:~/grpc# ./ni_grpc_device_server
Server listening on port 31763
Security is configured with insecure credentials.
^C
admin@NI-PXIe-8880-03095FC1:~/grpc# ./ni_grpc_device_server foo
Failed to parse config: The server configuration file was not found at: foo
```

syslog invocation with sad/happy paths:
```
admin@NI-PXIe-8880-03095FC1:~/grpc# ./ni_grpc_device_server --use-syslog
^C
admin@NI-PXIe-8880-03095FC1:~/grpc# ./ni_grpc_device_server --use-syslog foo
admin@NI-PXIe-8880-03095FC1:~/grpc# tail /var/log/messages
...
2021-05-21T19:16:47.096+00:00 NI-PXIe-8880-03095FC1 ni_grpc_server[31221]: Server listening on port 31763
2021-05-21T19:16:47.096+00:00 NI-PXIe-8880-03095FC1 ni_grpc_server[31221]: Security is configured with insecure credentials.
2021-05-21T19:16:54.260+00:00 NI-PXIe-8880-03095FC1 ni_grpc_server[31255]: Failed to parse config: The server configuration file was not found at: foo
````

### CODE REVIEW

File: source/server/linux/syslog_logging.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#include "syslog_logging.h"
+
+#include <syslog.h>
+
+namespace nidevice_grpc {
+namespace Logging {
+
+static const char* syslog_identity = "ni_grpc_server";
```
Comment (zhindes) [inline source/server/linux/syslog_logging.cpp:8]: that name is wrong, it should be "ni_grpc_device_server"

File: CMakeLists.txt [hunk 2]
```diff
@@ -176,9 +176,18 @@ foreach(api ${nidrivers})
   endif()
 endforeach()
 
+if(CMAKE_SYSTEM_NAME STREQUAL Linux)
+  set(SOURCE_OS
```
Comment (jonlugoatx) [inline CMakeLists.txt:180]: How about `OSDEP_SOURCE`? Without looking at where this is used and what it's set to, I would expect a variable named `SOURCE_OS` to contain the name of an operating system.

File: source/server/core_server.cpp [hunk 3]
```diff
@@ -87,25 +96,74 @@ static void RunServer(const std::string& config_file_path)
...
+{
+  Options options;
+  for (int i = 1; i < argc; ++i) {
+ #if defined(__GNUC__)
+    if (strcmp("--use-syslog", argv[i]) == 0) {
+      options.use_syslog = true;
+    }
+    else
+ #endif
+    if (strcmp("--help", argv[i]) == 0) {
```
Comment (jonlugoatx) [inline source/server/core_server.cpp:141]: Add `-h` too.

**Review by jonlugoatx — APPROVED** [review]

File: CMakeLists.txt [hunk 4]
```diff
@@ -176,9 +176,18 @@ foreach(api ${nidrivers})
   endif()
 endforeach()
 
+if(CMAKE_SYSTEM_NAME STREQUAL Linux)
+  set(OSDEP_SOURCE
+    "source/server/linux/syslog_logging.cpp")
+else()
+  set(OSDEP_SOURCE "")
+endif()
```
Comment (SparkingSpork) [inline CMakeLists.txt:184]: I think an alternative would be to add the source conditionally with[ target_sources](https://cmake.org/cmake/help/latest/command/target_sources.html), but I've never tried that so feel free to leave as-is for now.
```
if(CMAKE_SYSTEM_NAME STREQUAL Linux)
   target_sources(ni_grpc_device_server
      PRIVATE "source/server/linux/syslog_logging.cpp")
endif()
```

**Review by SparkingSpork — APPROVED** [review]

File: source/server/linux/syslog_logging.cpp [hunk 5]
```diff
@@ -3,7 +3,8 @@
 #include <syslog.h>
 
 namespace nidevice_grpc {
-namespace Logging {
+
```
Comment (jonlugoatx) [inline source/server/linux/syslog_logging.cpp:6]: We tend not to have blank lines between our namespaces. Same with the `}`s at the end.

**Review by jonlugoatx — APPROVED** [review]

**Review by tkrebes — APPROVED** [review]

File: CMakeLists.txt (hunk 4 above)
Comment (zhindes) [reply CMakeLists.txt]: That worked great, thanks! I'm new at CMake, so it's been interesting to try it out.

File: CMakeLists.txt [hunk 6]
```diff
@@ -177,8 +177,10 @@ foreach(api ${nidrivers})
 endforeach()
 
 add_executable(ni_grpc_device_server
+   ${OSDEP_SOURCE}
```
Comment (SparkingSpork) [inline CMakeLists.txt:180]: You shouldn't need this anymore.

File: CMakeLists.txt (hunk 4 above)
Comment (SparkingSpork) [reply CMakeLists.txt]: Glad it worked! I'm definitely still learning about it as we've refined things here.

**Pull request merged**